src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs

```diff
@@ -71,18 +63,30 @@ public override async Task<Compilation> InvokeAsync(Compilation oldCompilation,
                 }
             }
 
-            private class ProjectParseOptionsAction : SimpleCompilationTranslationAction<ProjectState>
+            internal class ReplaceAllSyntaxTreesAction : CompilationTranslationAction
             {
-                private static readonly Func<Compilation, ProjectState, CancellationToken, Task<Compilation>> s_action =
-                    (o, d, c) => Task.Run(() => ReplaceSyntaxTreesWithTreesFromNewProjectStateAsync(o, d, c), c);
+                private readonly ProjectState _state;
+
+                public ReplaceAllSyntaxTreesAction(ProjectState state)
+                {
+                    _state = state;
+                }
 
-                public ProjectParseOptionsAction(ProjectState state)
-                    : base(state, s_action)
+                public override async Task<Compilation> InvokeAsync(Compilation oldCompilation, CancellationToken cancellationToken)
                 {
+                    var syntaxTrees = new List<SyntaxTree>(capacity: _state.DocumentIds.Count);
```

Comment:
could this just be a literal array?

Reply:
It does, but then I have to track the array index. I was lazy, and given how expensive the operation is not sure it matters.